### DISCUSSION TITLE
QuestPDF + Native AOT + Rust

### DISCUSSION THREAD

**Discussion by ericsink** [QuestPDF/QuestPDF#496] [Show and tell]
Just for fun:

Demo video of calling QuestPDF from Rust using Native AOT and generated bindings:

https://youtu.be/XQI1SvvqbGk

Here's the blog entry which explains a bit more:

https://ericsink.com/native_aot/looking_ahead_questpdf_rust.html

**Comment by MarcinZiabek** [comment]
Apologies for the late reply - and thank you for sharing this! 

Your presentation and approach are genuinely interesting.

About three years ago, your work was one of the inspirations for exploring ports of QuestPDF to other platforms. Now that .NET Native AOT is finally stable and production-ready, we're actively exploring this direction. The most likely candidates are Kotlin/Java and TypeScript, with Python, PHP, and Go as possibilities - the final decision will depend on developer experience, platform popularity, effort per port, and audience size.

Your experiment also highlights something we've been thinking about: QuestPDF's fluent-first API doesn't translate naturally to every language, and Rust is a good example where it becomes quite verbose. Combined with the fact that Rust isn't a typical choice for backend document generation, it's probably not the first platform we'd target, but it's fascinating to see it working end-to-end.

We've already built a couple of internal prototypes using a similar approach, so it's great to see someone independently validating the idea. Would you like to share your experiences and know-how beyond your blog posts? Have you observed any usage of QuestPDF in Rust?

Thanks again for sharing! ❤️
